File: src/registry.py
"""Parse `.claude-plugin/marketplace.json` and resolve plugin sources.

The marketplace.json schema (per claude-code) lists plugins like:
    {
      "name": "vpailt-marketplace",
      "plugins": [
        {
          "name": "afv-library",
          "version": "1.7.5",
          "description": "...",
          "source": { "source": "url", "url": "https://github.com/vpailt/afv-library.git", "ref": "1.7.5-claude.1" }
        }
      ]
    }

A marketplace is therefore an *index* — each plugin entry tells us where to
fetch the actual plugin from.
"""
from __future__ import annotations

import json
import re
from typing import Optional

from .models import Plugin, PluginSource, InstallState


_GITHUB_URL_RE = re.compile(
    r"^(?:https?://(?:www\.)?github\.com/|git@github\.com:|github\.com/)"
    r"(?P<owner>[^/\s]+)/(?P<repo>[^/\s]+?)(?:\.git)?"
    r"(?:/(?:tree|blob|commits?|releases|pulls?|issues|wiki|actions)(?:/[^\s]*)?)?"
    r"/?(?:\?[^\s]*)?(?:#[^\s]*)?$"
)


def parse_marketplace_json(text: str, marketplace_name: str = "") -> list[Plugin]:
    """Parse a marketplace.json text into a list of Plugin entries.

    Each Plugin gets `latest_version` and `source` populated from the registry.
    Skills are NOT populated here — the caller fetches them on demand.
    """
    try:
        data = json.loads(text)
    except json.JSONDecodeError:
        return []
    if not isinstance(data, dict):
        return []
    name = marketplace_name or data.get("name") or ""
    raw_plugins = data.get("plugins") or []
    plugins: list[Plugin] = []
    for entry in raw_plugins:
        if not isinstance(entry, dict):
            continue
        src = _parse_source(entry.get("source"))
        plugins.append(Plugin(
            name=entry.get("name") or "",
            marketplace_name=name,
            latest_version=entry.get("version"),
            description=entry.get("description", "") or "",
            manifest=None,
            source=src,
            install_state=InstallState.NOT_INSTALLED,
            remote_present=True,
        ))
    return plugins


def _parse_source(raw) -> PluginSource:
    # The marketplace.json schema allows the plugin's `source` to be either a
    # dict (typed source: github / git-subdir / url) or a string. A string source
    # is a path relative to the marketplace repo itself (e.g. "./plugins/foo").
    if isinstance(raw, str):
        return PluginSource(kind="path", repo="", url="", ref="", path=raw)
    if not isinstance(raw, dict):
        return PluginSource(kind="", repo="", url="", ref="", path="")

    kind = raw.get("source", "") or ""
    url = raw.get("url", "") or ""
    repo = raw.get("repo", "") or ""
    ref = raw.get("ref", "") or ""
    path = raw.get("path", "") or ""

    if not repo and url:
        m = _GITHUB_URL_RE.match(url.strip())
        if m:
            repo = f"{m.group('owner')}/{m.group('repo')}"

    return PluginSource(kind=kind, repo=repo, url=url, ref=ref, path=path)


def parse_github_marketplace_url(url: str) -> Optional[str]:
    """Convert a github clone URL ('https://github.com/owner/repo.git') to 'owner/repo'."""
    if not url:
        return None
    m = _GITHUB_URL_RE.match(url.strip())
    return f"{m.group('owner')}/{m.group('repo')}" if m else None


def read_git_remote_origin(repo_dir) -> Optional[str]:
    """Read `.git/config` and return the origin URL — no `git` binary needed."""
    from pathlib import Path
    cfg = Path(repo_dir) / ".git" / "config"
    if not cfg.exists():
        return None
    try:
        text = cfg.read_text(encoding="utf-8", errors="replace")
    except Exception:
        return None
    in_origin = False
    for line in text.splitlines():
        s = line.strip()
        if s.startswith("[") and s.endswith("]"):
            in_origin = s == '[remote "origin"]'
            continue
        if in_origin and s.startswith("url"):
            _, _, value = s.partition("=")
            return value.strip() or None
    return None

File: src-tauri/src/registry.rs
//! Parse `.claude-plugin/marketplace.json` — port of src/registry.py.

use crate::models::{InstallState, Plugin, PluginSource};
use serde_json::Value;
use std::path::Path;

/// Extract `owner/repo` from a clone/web URL on **any** host (github.com or a
/// self-hosted Gitea instance like `https://git.almaviacx.local`).
///
/// Handles `https://host/owner/repo[.git][/tree/…][?…][#…]` and the SSH form
/// `git@host:owner/repo[.git]`. Takes the first two path segments as
/// owner/repo, which is correct for both forges' web and clone URLs.
pub fn parse_repo_from_url(url: &str) -> Option<String> {
    let url = url.trim();
    if url.is_empty() {
        return None;
    }
    // SSH form: git@host:owner/repo(.git)
    if let Some(rest) = url.strip_prefix("git@") {
        if let Some((_host, path)) = rest.split_once(':') {
            return owner_repo_from_path(path);
        }
    }
    let no_scheme = url
        .strip_prefix("https://")
        .or_else(|| url.strip_prefix("http://"))
        .unwrap_or(url);
    let no_scheme = no_scheme.strip_prefix("www.").unwrap_or(no_scheme);
    // Drop the host; keep the path.
    let (_host, path) = no_scheme.split_once('/')?;
    owner_repo_from_path(path)
}

fn owner_repo_from_path(path: &str) -> Option<String> {
    let path = path.split(['?', '#']).next().unwrap_or(path);
    let segs: Vec<&str> = path.split('/').filter(|s| !s.is_empty()).collect();
    if segs.len() < 2 {
        return None;
    }
    let owner = segs[0];
    let repo = segs[1].trim_end_matches(".git");
    if owner.is_empty() || repo.is_empty() {
        return None;
    }
    Some(format!("{owner}/{repo}"))
}

pub fn parse_marketplace_json(text: &str, marketplace_name: &str) -> Vec<Plugin> {
    let data: Value = match serde_json::from_str(text) {
        Ok(v) => v,
        Err(_) => return Vec::new(),
    };
    let Some(obj) = data.as_object() else {
        return Vec::new();
    };
    let name = if !marketplace_name.is_empty() {
        marketplace_name.to_string()
    } else {
        obj.get("name")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string()
    };
    let raw_plugins = obj
        .get("plugins")
        .and_then(|v| v.as_array())
        .cloned()
        .unwrap_or_default();
    let mut out = Vec::new();
    for entry in raw_plugins {
        let Some(o) = entry.as_object() else { continue };
        let src = parse_source(o.get("source").cloned().unwrap_or(Value::Null));
        out.push(Plugin {
            name: o
                .get("name")
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .to_string(),
            marketplace_name: name.clone(),
            latest_version: o
                .get("version")
                .and_then(|v| v.as_str())
                .map(String::from),
            description: o
                .get("description")
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .to_string(),
            manifest: None,
            source: Some(src),
            install_state: InstallState::NotInstalled,
            remote_present: true,
            ..Default::default()
        });
    }
    out
}

fn parse_source(raw: Value) -> PluginSource {
    if let Some(s) = raw.as_str() {
        return PluginSource {
            kind: "path".to_string(),
            path: s.to_string(),
            ..Default::default()
        };
    }
    let Some(obj) = raw.as_object() else {
        return PluginSource::default();
    };
    let kind = obj
        .get("source")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string();
    let url = obj
        .get("url")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string();
    let mut repo = obj
        .get("repo")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string();
    let r#ref = obj
        .get("ref")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string();
    let path = obj
        .get("path")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string();
    if repo.is_empty() && !url.is_empty() {
        if let Some(parsed) = parse_repo_from_url(&url) {
            repo = parsed;
        }
    }
    PluginSource {
        kind,
        repo,
        url,
        r#ref,
        path,
    }
}

/// Back-compat alias. Historically GitHub-only; now resolves owner/repo from
/// any forge host via [`parse_repo_from_url`]. Kept under the old name because
/// admin/command call sites import it.
pub fn parse_github_marketplace_url(url: &str) -> Option<String> {
    parse_repo_from_url(url)
}

pub fn read_git_remote_origin(repo_dir: &Path) -> Option<String> {
    let cfg = repo_dir.join(".git/config");
    let text = std::fs::read_to_string(cfg).ok()?;
    let mut in_origin = false;
    for line in text.lines() {
        let s = line.trim();
        if s.starts_with('[') && s.ends_with(']') {
            in_origin = s == "[remote \"origin\"]";
            continue;
        }
        if in_origin && s.starts_with("url") {
            if let Some((_, v)) = s.split_once('=') {
                let v = v.trim();
                if !v.is_empty() {
                    return Some(v.to_string());
                }
            }
        }
    }
    None
}
